soroban-env-host/src/vm/parsed_module.rs

```diff
@@ -0,0 +1,550 @@
+            // of "allowing uploads of a future-protocol contract that will go
+            // live as soon as the network upgrades to it") but there's a risk
+            // that the "future" protocol semantics baked in to a contract
+            // differ from the final semantics chosen by the network, so to be
+            // conservative we avoid even allowing this.
+            return Err(err!(
+                host,
+                (ScErrorType::WasmVm, ScErrorCode::InvalidInput),
+                "contract protocol number is newer than host",
+                got_proto
+            ));
+        }
+        Ok(())
+    }
+
+    fn module_custom_section(m: &Module, name: impl AsRef<str>) -> Option<&[u8]> {
+        m.custom_sections().iter().find_map(|s| {
+            if &*s.name == name.as_ref() {
+                Some(&*s.data)
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Returns the raw bytes content of a named custom section from the WASM
+    /// module loaded into the [Vm], or `None` if no such custom section exists.
+    #[allow(dead_code)]
+    pub fn custom_section(&self, name: impl AsRef<str>) -> Option<&[u8]> {
+        Self::module_custom_section(&self.module, name)
+    }
+
+    fn check_meta_section(host: &Host, m: &Module) -> Result<u64, HostError> {
+        if let Some(env_meta) = Self::module_custom_section(m, meta::ENV_META_V0_SECTION_NAME) {
+            let mut limits = DEFAULT_XDR_RW_LIMITS;
+            limits.len = env_meta.len();
+            let mut cursor = Limited::new(Cursor::new(env_meta), limits);
+            if let Some(env_meta_entry) = ScEnvMetaEntry::read_xdr_iter(&mut cursor).next() {
+                let ScEnvMetaEntry::ScEnvMetaKindInterfaceVersion(v) =
+                    host.map_err(env_meta_entry)?;
+                Self::check_contract_interface_version(host, v)?;
+                Ok(v)
+            } else {
+                Err(host.err(
+                    ScErrorType::WasmVm,
+                    ScErrorCode::InvalidInput,
+                    "contract missing environment interface version",
+                    &[],
+                ))
+            }
+        } else {
+            Err(host.err(
+                ScErrorType::WasmVm,
+                ScErrorCode::InvalidInput,
+                "contract missing metadata section",
+                &[],
+            ))
+        }
+    }
+
+    fn check_max_args(host: &Host, m: &Module) -> Result<(), HostError> {
+        for e in m.exports() {
+            match e.ty() {
+                wasmi::ExternType::Func(f) => {
+                    if f.params().len() > MAX_VM_ARGS || f.results().len() > MAX_VM_ARGS {
+                        return Err(host.err(
+                            ScErrorType::WasmVm,
+                            ScErrorCode::InvalidInput,
+                            "Too many arguments or results in wasm export",
+                            &[],
```

Comment:
Can we include the actual values into the error message? Also applies to most of the errors here (where possible)

Reply:
Fixed (a fair amount -- always room for improvement)